controller/volume_clone_controller.go

```diff
@@ -0,0 +1,268 @@
+
+	vcc.logger.Infof("Start Longhorn volume clone controller")
+	defer vcc.logger.Infof("Shutting down Longhorn volume clone controller")
+
+	if !cache.WaitForNamedCacheSync(vcc.name, stopCh, vcc.cacheSyncs...) {
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(vcc.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+}
+
+func (vcc *VolumeCloneController) worker() {
+	for vcc.processNextWorkItem() {
+	}
+}
+
+func (vcc *VolumeCloneController) processNextWorkItem() bool {
+	key, quit := vcc.queue.Get()
+	if quit {
+		return false
+	}
+	defer vcc.queue.Done(key)
+	err := vcc.syncHandler(key.(string))
+	vcc.handleErr(err, key)
+	return true
+}
+
+func (vcc *VolumeCloneController) handleErr(err error, key interface{}) {
+	if err == nil {
+		vcc.queue.Forget(key)
+		return
+	}
+
+	vcc.logger.WithError(err).Warnf("Error syncing Longhorn volume %v", key)
+	vcc.queue.AddRateLimited(key)
+	return
+}
+
+func (vcc *VolumeCloneController) syncHandler(key string) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "%v: failed to sync volume %v", vcc.name, key)
+	}()
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	if namespace != vcc.namespace {
+		return nil
+	}
+	return vcc.reconcile(name)
+}
+
+func (vcc *VolumeCloneController) reconcile(volName string) (err error) {
+	vol, err := vcc.ds.GetVolume(volName)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+		return nil
+	}
+
+	if !vcc.isResponsibleFor(vol) {
+		return nil
+	}
+
+	va, err := vcc.ds.GetLHVolumeAttachment(types.GetLHVolumeAttachmentNameFromVolumeName(volName))
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+		vcc.enqueueVolumeAfter(vol, constant.LonghornVolumeAttachmentNotFoundRetryPeriod)
+		return nil
+	}
+	existingVA := va.DeepCopy()
+	defer func() {
+		if err != nil {
+			return
+		}
+		if reflect.DeepEqual(existingVA.Spec, va.Spec) {
+			return
+		}
+
+		if _, err = vcc.ds.UpdateLHVolumeAttachment(va); err != nil {
+			return
+		}
+	}()
+
+	expectedAttachmentTickets := make(map[string]bool)
+
+	// case 1: this volume is target of a clone
+	if isTargetVolumeOfCloning(vol) {
+		cloningAttachmentTicketID := longhorn.GetAttachmentTicketID(longhorn.AttacherTypeVolumeCloneController, volName)
+		vcc.createOrUpdateAttachmentTicket(cloningAttachmentTicketID, vol.Status.OwnerID, va, longhorn.TrueValue)
+		expectedAttachmentTickets[cloningAttachmentTicketID] = true
+	}
+
+	// case 2: this volume is source of a clone
+	vols, err := vcc.ds.ListVolumes()
+	if err != nil {
+		return err
+	}
+	for _, v := range vols {
+		attachmentTicketID := longhorn.GetAttachmentTicketID(longhorn.AttacherTypeVolumeCloneController, v.Name)
+		if isTargetVolumeOfCloning(v) && types.GetVolumeName(v.Spec.DataSource) == vol.Name {
+			vcc.createOrUpdateAttachmentTicket(attachmentTicketID, vol.Status.OwnerID, va, longhorn.AnyValue)
+			expectedAttachmentTickets[attachmentTicketID] = true
+		}
+	}
```

Comment:
I'm a little confused about case 2. When the `targetVolume.Spec.DataSource` is updated, the reconciliation be triggered and then creates an attachmentTicket. Do we need a extra iteration here?

Reply:
The ticket you are mentioning is for the target volume, right?

We also need to attach the source volume for the cloning. This is the reason for the extra iteration.